### PR TITLE
Changed CMAKE_CXX_FLAGS to c++17, add target links to Abseil library and c_str() addition when logging

### PR DESCRIPTION
**Pull request by fordth**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by jamolina** [comment]
Thank you for your contribution! We will review your changes and get back to you as soon as possible

**Comment by smhmhmd** [comment]
@fordth I have added one more PR on top of yours - https://github.com/fordth/credentials-fetcher/pull/2

**Comment by smhmhmd** [comment]
@saikiranakula-amzn - Please review PR

**Comment by fordth** [comment]
@smhmhmd I just tried running cmake -DCMAKE_CXX_STANDARD=17 ../ on Ubuntu 20.04 and have now since received a number of errors. I will follow up with an output of the errors today ,but performing a cmake on the PR will repo the issue in Ubuntu

**Comment by smhmhmd** [comment]
> @smhmhmd I just tried running cmake -DCMAKE_CXX_STANDARD=17 ../ on Ubuntu 20.04 and have now since received a number of errors. I will follow up with an output of the errors today ,but performing a cmake on the PR will repo the issue in Ubuntu

Hi @fordth ,
You dont need to run  cmake -DCMAKE_CXX_STANDARD=17 ../., cmake ../ is sufficient if you already have gRPC installed.
To install gRPC and krb5, you need to run cmake in the dependencies directory.
I have the steps in the comment of the PR - https://github.com/fordth/credentials-fetcher/pull/2

I just retested in my instance.

**Comment by smhmhmd** [comment]
@fordth, I have some time till the end of the hour if you want to meet or we can meet on Monday.

**Comment by smhmhmd** [comment]
@fordth
If you are able to test, I will get your PR merged.

**Comment by fordth** [comment]
> 

We plan to test the compilation morning of Wednesday 9/13. Once this PR is then completed, I will create a new PR containing the additional changes we spoke about.  I will provide an update tomorrow

File: docker/Dockerfile-ubuntu-22.04 [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    && mkdir build \
+    && cd build \
+    && cmake -DgRPC_INSTALL=ON -DgRPC_BUILD_TESTS=OFF -DCMAKE_CXX_STANDARD=17 ../ \
+    && make -j4 \
+    && make install
+
+RUN apt-get update \
+    && apt-get install -y dotnet-sdk-6.0
+
+RUN cd /root && git clone https://github.com/fordth/credentials-fetcher \
```
Comment (ramapullareddyc) [inline docker/Dockerfile-ubuntu-22.04:20]: Should we change the URL to point to correct branch?

**Comment by fordth** [comment]
Good catch, I updated the PR

File: docker/Dockerfile-ubuntu-20.04 [hunk 2]
```diff
@@ -0,0 +1,39 @@
+FROM ubuntu:20.04 as base
+
+RUN apt-get update \
+    && DEBIAN_FRONTEND="noninteractive" TZ="America/New_York" \
```
Comment (ramapullareddyc) [inline docker/Dockerfile-ubuntu-20.04]: UTC time zone is better and applicable universally and/or passing in a time zone would also be good. Same for 22.04 docker file as well.

File: docker/Dockerfile-ubuntu-20.04 [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    && make install
+
+RUN apt-get install wget -y && wget https://packages.microsoft.com/config/ubuntu/20.04/packages-microsoft-prod.deb -O packages-microsoft-prod.deb \
+    && dpkg -i packages-microsoft-prod.deb \
+    && rm packages-microsoft-prod.deb \
+    && apt-get update \
+    && apt-get install -y dotnet-sdk-6.0 \
+    && ln -s '/usr/share/dotnet' '/usr/lib/dotnet'
+
+RUN cd /root && git clone https://github.com/fordth/credentials-fetcher \
```
Comment (ramapullareddyc) [inline docker/Dockerfile-ubuntu-20.04]: When do we plan to update URL to point to official repository? Same for 22.04 docker file as well.

File: docker/Dockerfile-ubuntu-22.04 [hunk 4]
```diff
@@ -0,0 +1,31 @@
+FROM ubuntu:22.04 as base
+
+RUN DEBIAN_FRONTEND="noninteractive" TZ="America/New_York" \
```
Comment (ramapullareddyc) [inline docker/Dockerfile-ubuntu-22.04]: Do we not need to build cMake and grpc like how we are doing in ubuntu-20.04 docker file?

File: docker/Dockerfile-ubuntu-20.04 (hunk 2 above)
Comment (smhmhmd) [reply docker/Dockerfile-ubuntu-20.04]: Yes, an environment variable can be passed into Docker using --env option.

File: docker/Dockerfile-ubuntu-22.04 (hunk 4 above)
Comment (smhmhmd) [reply docker/Dockerfile-ubuntu-22.04]: The Dockerfile for Ubuntu-22.04 needs to be updated with changes from the Dockerfile for Ubuntu-20.04

File: docker/Dockerfile-ubuntu-20.04 (hunk 3 above)
Comment (fordth) [reply docker/Dockerfile-ubuntu-20.04]: The PR has been updated with the correct repository

File: docker/Dockerfile-ubuntu-22.04 (hunk 4 above)
Comment (fordth) [reply docker/Dockerfile-ubuntu-22.04]: I can work on the ubuntu 22.04 Docker file but I may need a day or two to complete it.

File: CMakeLists.txt [hunk 5]
```diff
@@ -126,15 +133,18 @@ add_custom_command(
...
+        PUBLIC systemd boost_program_options krb5 glib-2.0 cf_gmsa_service_private
+        boost_filesystem boost_system crypto protobuf kadm5srv_mit kdb5 gssrpc gssapi_krb5 gssrpc k5crypto com_err krb5support resolv utf8_validity)
+else()
 target_link_libraries(credentials-fetcherd
         PUBLIC systemd boost_program_options krb5 glib-2.0 cf_gmsa_service_private
         boost_filesystem boost_system crypto
         kadm5srv_mit kdb5 gssrpc gssapi_krb5 gssrpc k5crypto
         com_err krb5support resolv)
-
-target_link_libraries(credentials-fetcherd
```
Comment (ramapullareddyc) [inline CMakeLists.txt:135]: Do we need to remove this line for non-ubuntu distros? "api/CMakeLists.txt" file links using same variables for non-ubuntu code path.

Comment (smhmhmd) [reply CMakeLists.txt:135]: +1, kerberos libs are duplicated, this can be cleaned up.

Comment (fordth) [reply CMakeLists.txt:135]: My latest update in the PR now includes the updated Ubuntu 22.04 docker file.

File: docker/Dockerfile-ubuntu-20.04 (hunk 2 above)
Comment (fordth) [reply docker/Dockerfile-ubuntu-20.04]: The PR has been updated to require --build-arg TIME_ZONE during docker build

**Review by saikiranakula-amzn — APPROVED** [review]
testing on fedora

**Review by ramapullareddyc — APPROVED** [review]

**Pull request merged**